### PR TITLE
Add option to support criu manage cgroups mode for dump and restore

### PR DESCRIPTION
**Pull request by huikang**
Signed-off-by: Hui Kang <hkang.sunysb@gmail.com>

### CODE REVIEW

**Comment by boucher** [comment]
Will this fail unless you're running the most recent Criu?

**Comment by huikang** [comment]
@boucher you are right. It depends on a recent criu patch (https://github.com/xemul/criu/commit/3d1e8c688347a1b0ed062197eb4e640f85f75660)
So I think we should hold this until next release of CRIU.

**Comment by avagin** [comment]
We don't break backward compatibility. It will work fine if you don't use the manage-cgroups-mode option. And you can see that the continuous validation passes for this pull request.

**Comment by boucher** [comment]
Should we make a note somewhere what version is required to use this
feature? Everything else will work with the minimum required criu version
right now I think.

**Comment by huikang** [comment]
@boucher @avagin If we want to add a note as criu version regarding when to enable the manage mode, is it 1.6 as the minimum requirement?

**Comment by dqminh** [comment]
> Should we make a note somewhere what version is required to use this
feature? Everything else will work with the minimum required criu version
right now I think.

I think so. It would also be nice to detect underlying criu version and display a nice error message when this feature is used on non-supported criu.

File: libcontainer/container_linux.go [hunk 1]
```diff
@@ -376,6 +383,14 @@ func (c *linuxContainer) Checkpoint(criuOpts *CriuOpts) error {
 		}
 	}
 
+	// append optional manage cgroups mode
+	if criuOpts.ManageCgroupsMode != 0 {
+		if err := c.checkCriuVersion("1.6"); err != nil {
+			return err
```
Comment (huikang) [inline libcontainer/container_linux.go]: @dqminh we check the criu version when using this feature. Thanks.

File: libcontainer/criu_opts.go [hunk 2]
```diff
@@ -1,5 +1,14 @@
 package libcontainer
 
+type cg_mode uint32
+
+const (
+	CRIU_CG_MODE_SOFT cg_mode = 3 + iota
+	CRIU_CG_MODE_FULL
+	CRIU_CG_MODE_STRICT
+	CRIU_CG_MODE_DEFAULT
+)
```
Comment (huikang) [inline libcontainer/criu_opts.go]: @avagin Constants are defined here according to Ruslan's criu patch. The number begins with 3; as restoring a container requires cgroup. Thanks.

Comment (thaJeztah) [inline libcontainer/criu_opts.go]: (just passing by); if that needs explanation - would it be useful to add that information as a comment in the code?

Comment (huikang) [inline libcontainer/criu_opts.go]: Hi, Sebastiaan,
Thanks for your suggestions. I have added comment to these constants.

File: libcontainer/container_linux.go [hunk 3]
```diff
@@ -376,6 +383,14 @@ func (c *linuxContainer) Checkpoint(criuOpts *CriuOpts) error {
 		}
 	}
 
+	// append optional manage cgroups mode
+	if criuOpts.ManageCgroupsMode != 0 {
+		if err := c.checkCriuVersion("1.6"); err != nil {
```
Comment (avagin) [inline libcontainer/container_linux.go]: Your patch for criu was commited after 1.6.
```
# git describe 3d1e8c688347a1b0ed062197eb4e640f85f75660
v1.6-164-g3d1e8c6
```

Comment (avagin) [inline libcontainer/container_linux.go]: @xemul what is the right way to check criu version if it was compiled from git?

Comment (xemul) [inline libcontainer/container_linux.go]: @avagin , "criu --version"'s line GitID: contains all the necessary info ;)

Comment (avagin) [inline libcontainer/container_linux.go]: The dash symbol in a criu version means that it was compiled from git. So if we found dash we can increment the last number in version.
For example:
1.6-smth should be converted into 1.7
1.6.2-smth should be converted int 1.6.3

Comment (huikang) [inline libcontainer/container_linux.go]: @avagin @xemul Thanks for the instructions. So the pseudo code would be
output := "criu --version"
if output containers "GitID"
    // use the line GitID: 1.6.2-smth or 1.6-smth
    // increment the number by one
else
    // use the line Version: 1.6 as in the exisitng code

Correct?

Comment (avagin) [inline libcontainer/container_linux.go]: Yes

**Comment by huikang** [comment]
@avagin could you help to take a look at the patch? Any other suggestions? Thanks.

File: libcontainer/container_linux.go (hunk 3 above)
Comment (avagin) [inline libcontainer/container_linux.go]: Are you going to fix this issue?

Comment (huikang) [inline libcontainer/container_linux.go]: @avagin I will change it to 1.7. Sorry about the delay.

**Comment by huikang** [comment]
@avagin rebased to the current HEAD to make Jenkins happy. Thanks.

**Comment by avagin** [comment]
LGTM

I recommend to re-work commits. It isn't interesting to read history how a feature were developed.

**Comment by huikang** [comment]
@avagin The commits are squashed. Thanks.

**Comment by crosbymichael** [comment]
LGTM

ping @mrunalp @LK4D4

**Comment by huikang** [comment]
@mrunalp @LK4D4 rebased to the master branch. Thanks.

**Comment by LK4D4** [comment]
LGTM

**Pull request merged**